File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        int i=0;
        int middle=0;
        ListNode *temp=head;
        ListNode *curr=NULL;
        if(head->next==NULL)
        {
            return NULL;
        }
        for(i=0;temp!=NULL;i++)
        {
            temp=temp->next;
        }
        middle=i/2;
        temp=head;
        for(i=0;i<middle-1;i++)
        {
            temp=temp->next;
        }
        curr=temp->next;
        temp->next=temp->next->next;
        curr->next=NULL;
        delete curr;
        return head;
    }
};

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = head
        slow = head
        fast = head
        if head.next == None:
            return None
        while fast and fast.next:
            prev = slow
            slow  = slow.next
            fast = fast.next.next

        temp=slow
        prev.next = slow.next
        temp.next=None
        return head
        # if head.next==None:
        #     return None
        
        # slow = head
        # fast = head
        
        # while fast and fast.next:
        #     fast = fast.next.next
        #     if fast and fast.next:
        #         slow = slow.next
        
        # slow.next = slow.next.next
        
        # return head
    
    # def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
    #     turtle, hare = head, head
    #     while hare.next and hare.next.next and hare.next.next.next != None:
    #         turtle = turtle.next
    #         hare = hare.next.next
        
    #     if head.next:
    #         temp = turtle.next.next
    #         turtle.next.next = None
    #         turtle.next = temp
    #         return head

        